File: nanocode.js
#!/usr/bin/env node
/**
 * nanocode - minimal claude code alternative
 */

const { execSync, spawn } = require("child_process");
const fs = require("fs");
const path = require("path");
const readline = require("readline");

const OPENROUTER_KEY = process.env.OPENROUTER_API_KEY;
const API_URL = OPENROUTER_KEY
  ? "https://openrouter.ai/api/v1/messages"
  : "https://api.anthropic.com/v1/messages";
const MODEL =
  process.env.MODEL ||
  (OPENROUTER_KEY ? "anthropic/claude-opus-4.5" : "claude-opus-4-5");

// ANSI colors
const RESET = "\x1b[0m",
  BOLD = "\x1b[1m",
  DIM = "\x1b[2m";
const BLUE = "\x1b[34m",
  CYAN = "\x1b[36m",
  GREEN = "\x1b[32m",
  YELLOW = "\x1b[33m",
  RED = "\x1b[31m";

// --- Tool implementations ---
function bash(args) {
  return new Promise((resolve) => {
    const proc = spawn(args.cmd, { shell: true });
    const outputLines = [];

    const handleData = (data) => {
      const lines = data.toString().split("\n");
      lines.forEach((line) => {
        if (line) {
          console.log(`  ${DIM}│ ${line}${RESET}`);
          outputLines.push(line);
        }
      });
    };

    proc.stdout.on("data", handleData);
    proc.stderr.on("data", handleData);

    const timeout = setTimeout(() => {
      proc.kill();
      outputLines.push("(timed out after 30s)");
      resolve(outputLines.join("\n") || "(empty)");
    }, 30000);

    proc.on("close", () => {
      clearTimeout(timeout);
      resolve(outputLines.join("\n").trim() || "(empty)");
    });
  });
}

function readFile(args) {
  const filePath = args.path;
  if (!fs.existsSync(filePath)) {
    return `error: file not found: ${filePath}`;
  }
  const content = fs.readFileSync(filePath, "utf-8");
  const lines = content.split("\n");
  const numbered = lines.map(
    (line, i) => `${String(i + 1).padStart(4)} | ${line}`
  );
  return numbered.join("\n");
}

function writeFile(args) {
  const filePath = args.path;
  const content = args.content;
  const dir = path.dirname(filePath);
  if (dir && dir !== ".") {
    fs.mkdirSync(dir, { recursive: true });
  }
  fs.writeFileSync(filePath, content);
  const lines = content.split("\n").length;
  return `wrote ${lines} lines to ${filePath}`;
}

function glob(args) {
  const pattern = args.pattern;
  const { globSync } = require("fs");
  // Use a simple recursive glob implementation
  const matches = globRecursive(pattern);
  return matches.length ? matches.sort().join("\n") : "(no matches)";
}

function globRecursive(pattern) {
  const results = [];
  const parts = pattern.split("/");
  const hasGlobstar = pattern.includes("**");

  function walkDir(dir, remainingParts) {
    if (remainingParts.length === 0) return;

    const currentPart = remainingParts[0];
    const isLast = remainingParts.length === 1;

    if (!fs.existsSync(dir) || !fs.statSync(dir).isDirectory()) return;

    const entries = fs.readdirSync(dir, { withFileTypes: true });

    if (currentPart === "**") {
      // Match zero or more directories
      walkDir(dir, remainingParts.slice(1));
      for (const entry of entries) {
        if (entry.isDirectory()) {
          walkDir(path.join(dir, entry.name), remainingParts);
        }
      }
    } else {
      const regex = new RegExp(
        "^" + currentPart.replace(/\*/g, ".*").replace(/\?/g, ".") + "$"
      );
      for (const entry of entries) {
        if (regex.test(entry.name)) {
          const fullPath = path.join(dir, entry.name);
          if (isLast) {
            results.push(fullPath);
          } else if (entry.isDirectory()) {
            walkDir(fullPath, remainingParts.slice(1));
          }
        }
      }
    }
  }

  const startDir = parts[0] === "" ? "/" : ".";
  const startParts = parts[0] === "" ? parts.slice(1) : parts;
  walkDir(startDir, startParts);
  return results;
}

// --- Tool definitions ---
const TOOLS = {
  bash: {
    description: "Run shell command",
    schema: { cmd: "string" },
    fn: bash,
  },
  read_file: {
    description: "Read file content with line numbers",
    schema: { path: "string" },
    fn: readFile,
  },
  write_file: {
    description: "Write content to file (creates dirs if needed)",
    schema: { path: "string", content: "string" },
    fn: writeFile,
  },
  glob: {
    description: "Find files matching pattern (supports **)",
    schema: { pattern: "string" },
    fn: glob,
  },
};

async function runTool(name, args) {
  try {
    const result = await TOOLS[name].fn(args);
    return result;
  } catch (err) {
    return `error: ${err.message}`;
  }
}

function makeSchema() {
  return Object.entries(TOOLS).map(([name, { description, schema }]) => {
    const properties = {};
    const required = [];
    for (const [paramName, paramType] of Object.entries(schema)) {
      const isOptional = paramType.endsWith("?");
      const baseType = paramType.replace(/\?$/, "");
      properties[paramName] = {
        type: baseType === "number" ? "integer" : baseType,
      };
      if (!isOptional) required.push(paramName);
    }
    return {
      name,
      description,
      input_schema: { type: "object", properties, required },
    };
  });
}

async function callApi(messages, systemPrompt) {
  const headers = {
    "Content-Type": "application/json",
    "anthropic-version": "2023-06-01",
  };

  if (OPENROUTER_KEY) {
    headers["Authorization"] = `Bearer ${OPENROUTER_KEY}`;
  } else {
    headers["x-api-key"] = process.env.ANTHROPIC_API_KEY || "";
  }

  const response = await fetch(API_URL, {
    method: "POST",
    headers,
    body: JSON.stringify({
      model: MODEL,
      max_tokens: 8192,
      system: systemPrompt,
      messages,
      tools: makeSchema(),
    }),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`API error: ${response.status} ${text}`);
  }

  return response.json();
}

function separator() {
  const cols = process.stdout.columns || 80;
  return `${DIM}${"─".repeat(Math.min(cols, 80))}${RESET}`;
}

function renderMarkdown(text) {
  return text.replace(/\*\*(.+?)\*\*/g, `${BOLD}$1${RESET}`);
}

async function main() {
  console.log(
    `${BOLD}nanocode${RESET} | ${DIM}${MODEL} (${OPENROUTER_KEY ? "OpenRouter" : "Anthropic"}) | ${process.cwd()}${RESET}\n`
  );

  const messages = [];
  const systemPrompt = `Concise coding assistant. cwd: ${process.cwd()}`;

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const prompt = () => {
    console.log(separator());
    rl.question(`${BOLD}${BLUE}❯${RESET} `, async (userInput) => {
      userInput = userInput.trim();
      console.log(separator());

      if (!userInput) {
        prompt();
        return;
      }

      if (userInput === "/q" || userInput === "exit") {
        rl.close();
        return;
      }

      if (userInput === "/c") {
        messages.length = 0;
        console.log(`${GREEN}⏺ Cleared conversation${RESET}`);
        prompt();
        return;
      }

      messages.push({ role: "user", content: userInput });

      try {
        // Agentic loop: keep calling API until no more tool calls
        while (true) {
          const response = await callApi(messages, systemPrompt);
          const contentBlocks = response.content || [];
          const toolResults = [];

          for (const block of contentBlocks) {
            if (block.type === "text") {
              console.log(`\n${CYAN}⏺${RESET} ${renderMarkdown(block.text)}`);
            }

            if (block.type === "tool_use") {
              const toolName = block.name;
              const toolArgs = block.input;
              const argPreview = String(Object.values(toolArgs)[0]).slice(
                0,
                50
              );
              console.log(
                `\n${GREEN}⏺ ${toolName.charAt(0).toUpperCase() + toolName.slice(1)}${RESET}(${DIM}${argPreview}${RESET})`
              );

              const result = await runTool(toolName, toolArgs);
              const resultLines = result.split("\n");
              let preview = resultLines[0].slice(0, 60);
              if (resultLines.length > 1) {
                preview += ` ... +${resultLines.length - 1} lines`;
              } else if (resultLines[0].length > 60) {
                preview += "...";
              }
              console.log(`  ${DIM}⎿  ${preview}${RESET}`);

              toolResults.push({
                type: "tool_result",
                tool_use_id: block.id,
                content: result,
              });
            }
          }

          messages.push({ role: "assistant", content: contentBlocks });

          if (toolResults.length === 0) break;
          messages.push({ role: "user", content: toolResults });
        }

        console.log();
      } catch (err) {
        console.log(`${RED}⏺ Error: ${err.message}${RESET}`);
      }

      prompt();
    });
  };

  rl.on("close", () => {
    console.log();
    process.exit(0);
  });

  prompt();
}

main();

File: nanocode.py
#!/usr/bin/env python3
"""nanocode - minimal claude code alternative"""

import glob as globlib
import json
import os
import re
import subprocess
import urllib.request
import urllib.parse

OPENROUTER_KEY = os.environ.get("OPENROUTER_API_KEY")
API_URL = (
    "https://openrouter.ai/api/v1/messages"
    if OPENROUTER_KEY
    else "https://api.anthropic.com/v1/messages"
)
MODEL = os.environ.get(
    "MODEL", "anthropic/claude-opus-4.5" if OPENROUTER_KEY else "claude-opus-4-5"
)

# ANSI colors
RESET, BOLD, DIM = "\033[0m", "\033[1m", "\033[2m"
BLUE, CYAN, GREEN, YELLOW, RED = (
    "\033[34m",
    "\033[36m",
    "\033[32m",
    "\033[33m",
    "\033[31m",
)


# --- Tool implementations ---
def bash(args):
    proc = subprocess.Popen(
        args["cmd"],
        shell=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
    )
    output_lines = []
    try:
        while True:
            line = proc.stdout.readline()
            if not line and proc.poll() is not None:
                break
            if line:
                print(f"  {DIM}│ {line.rstrip()}{RESET}", flush=True)
                output_lines.append(line)
        proc.wait(timeout=30)
    except subprocess.TimeoutExpired:
        proc.kill()
        output_lines.append("\n(timed out after 30s)")
    return "".join(output_lines).strip() or "(empty)"


def read_file(args):
    path = args["path"]
    if not os.path.exists(path):
        return f"error: file not found: {path}"
    with open(path, "r") as f:
        content = f.read()
    lines = content.split("\n")
    numbered = [f"{i+1:4} | {line}" for i, line in enumerate(lines)]
    return "\n".join(numbered)


def write_file(args):
    path = args["path"]
    content = args["content"]
    os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
    with open(path, "w") as f:
        f.write(content)
    lines = len(content.split("\n"))
    return f"wrote {lines} lines to {path}"


def glob(args):
    pattern = args["pattern"]
    matches = globlib.glob(pattern, recursive=True)
    return "\n".join(sorted(matches)) if matches else "(no matches)"


def web_search(args):
    query = args["query"]
    encoded_query = urllib.parse.quote(query)
    url = f"https://html.duckduckgo.com/html/?q={encoded_query}"
    
    req = urllib.request.Request(
        url,
        headers={
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        },
    )
    
    try:
        with urllib.request.urlopen(req, timeout=10) as response:
            html = response.read().decode("utf-8")
    except Exception as e:
        return f"search error: {e}"
    
    # Parse results from DuckDuckGo HTML
    results = []
    # Match result blocks: title, url, snippet
    pattern = r'<a[^>]*class="result__a"[^>]*href="([^"]*)"[^>]*>(.+?)</a>.*?<a[^>]*class="result__snippet"[^>]*>(.+?)</a>'
    matches = re.findall(pattern, html, re.DOTALL)
    
    for url, title, snippet in matches[:5]:  # Top 5 results
        # Clean HTML tags
        title = re.sub(r"<[^>]+>", "", title).strip()
        snippet = re.sub(r"<[^>]+>", "", snippet).strip()
        # Decode URL redirect
        if "uddg=" in url:
            url = urllib.parse.unquote(url.split("uddg=")[1].split("&")[0])
        results.append(f"**{title}**\n{url}\n{snippet}\n")
    
    return "\n".join(results) if results else "no results found"


def gh(args):
    """Execute GitHub CLI commands"""
    cmd = args["cmd"]
    # Ensure the command starts with 'gh'
    if not cmd.strip().startswith("gh "):
        cmd = f"gh {cmd}"
    
    proc = subprocess.Popen(
        cmd,
        shell=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
    )
    output_lines = []
    try:
        while True:
            line = proc.stdout.readline()
            if not line and proc.poll() is not None:
                break
            if line:
                print(f"  {DIM}│ {line.rstrip()}{RESET}", flush=True)
                output_lines.append(line)
        proc.wait(timeout=60)
    except subprocess.TimeoutExpired:
        proc.kill()
        output_lines.append("\n(timed out after 60s)")
    return "".join(output_lines).strip() or "(empty)"


# --- Tool definitions: (description, schema, function) ---

TOOLS = {
    "bash": (
        "Run shell command",
        {"cmd": "string"},
        bash,
    ),
    "read_file": (
        "Read file content with line numbers",
        {"path": "string"},
        read_file,
    ),
    "write_file": (
        "Write content to file (creates dirs if needed)",
        {"path": "string", "content": "string"},
        write_file,
    ),
    "glob": (
        "Find files matching pattern (supports **)",
        {"pattern": "string"},
        glob,
    ),
    "web_search": (
        "Search the web using DuckDuckGo, returns top results with titles, URLs and snippets",
        {"query": "string"},
        web_search,
    ),
    "gh": (
        "Execute GitHub CLI (gh) commands. Examples: 'gh repo view', 'gh issue list', 'gh pr create', 'gh pr list', 'gh release list'",
        {"cmd": "string"},
        gh,
    ),
}


def run_tool(name, args):
    try:
        return TOOLS[name][2](args)
    except Exception as err:
        return f"error: {err}"


def make_schema():
    result = []
    for name, (description, params, _fn) in TOOLS.items():
        properties = {}
        required = []
        for param_name, param_type in params.items():
            is_optional = param_type.endswith("?")
            base_type = param_type.rstrip("?")
            properties[param_name] = {
                "type": "integer" if base_type == "number" else base_type
            }
            if not is_optional:
                required.append(param_name)
        result.append(
            {
                "name": name,
                "description": description,
                "input_schema": {
                    "type": "object",
                    "properties": properties,
                    "required": required,
                },
            }
        )
    return result


def call_api(messages, system_prompt):
    request = urllib.request.Request(
        API_URL,
        data=json.dumps(
            {
                "model": MODEL,
                "max_tokens": 8192,
                "system": system_prompt,
                "messages": messages,
                "tools": make_schema(),
            }
        ).encode(),
        headers={
            "Content-Type": "application/json",
            "anthropic-version": "2023-06-01",
            **(
                {"Authorization": f"Bearer {OPENROUTER_KEY}"}
                if OPENROUTER_KEY
                else {"x-api-key": os.environ.get("ANTHROPIC_API_KEY", "")}
            ),
        },
    )
    response = urllib.request.urlopen(request)
    return json.loads(response.read())


def separator():
    return f"{DIM}{'─' * min(os.get_terminal_size().columns, 80)}{RESET}"


def render_markdown(text):
    return re.sub(r"\*\*(.+?)\*\*", f"{BOLD}\\1{RESET}", text)


def main():
    print(
        f"{BOLD}nanocode{RESET} | {DIM}{MODEL} ({'OpenRouter' if OPENROUTER_KEY else 'Anthropic'}) | {os.getcwd()}{RESET}\n"
    )
    messages = []
    system_prompt = f"Concise coding assistant. cwd: {os.getcwd()}"

    while True:
        try:
            print(separator())
            user_input = input(f"{BOLD}{BLUE}❯{RESET} ").strip()
            print(separator())
            if not user_input:
                continue
            if user_input in ("/q", "exit"):
                break
            if user_input == "/c":
                messages = []
                print(f"{GREEN}⏺ Cleared conversation{RESET}")
                continue

            messages.append({"role": "user", "content": user_input})

            # agentic loop: keep calling API until no more tool calls
            while True:
                response = call_api(messages, system_prompt)
                content_blocks = response.get("content", [])
                tool_results = []

                for block in content_blocks:
                    if block["type"] == "text":
                        print(f"\n{CYAN}⏺{RESET} {render_markdown(block['text'])}")

                    if block["type"] == "tool_use":
                        tool_name = block["name"]
                        tool_args = block["input"]
                        arg_preview = str(list(tool_args.values())[0])[:50]
                        print(
                            f"\n{GREEN}⏺ {tool_name.capitalize()}{RESET}({DIM}{arg_preview}{RESET})"
                        )

                        result = run_tool(tool_name, tool_args)
                        result_lines = result.split("\n")
                        preview = result_lines[0][:60]
                        if len(result_lines) > 1:
                            preview += f" ... +{len(result_lines) - 1} lines"
                        elif len(result_lines[0]) > 60:
                            preview += "..."
                        print(f"  {DIM}⎿  {preview}{RESET}")

                        tool_results.append(
                            {
                                "type": "tool_result",
                                "tool_use_id": block["id"],
                                "content": result,
                            }
                        )

                messages.append({"role": "assistant", "content": content_blocks})

                if not tool_results:
                    break
                messages.append({"role": "user", "content": tool_results})

            print()

        except (KeyboardInterrupt, EOFError):
            break
        except Exception as err:
            print(f"{RED}⏺ Error: {err}{RESET}")


if __name__ == "__main__":
    main()
